modules/runner/src/main/ts/runner/Run.ts

```diff
@@ -93,19 +94,25 @@ export const runWithTimeout = (runnable: Runnable, context: Context, defaultTime
     return run(runnable, context);
   } else {
     return new Promise((resolve, reject) => {
-      let timedOut = false;
-      const timer = setTimeout(() => {
-        timedOut = true;
-        reject(Failure.prepFailure(new Error(`Test ran too long - timeout of ${timeout}ms exceeded`)));
-      }, timeout);
+      const timer = Timer(() => {
+        cleanup();
```

Comment:
There is a lot calling things before it's really defined here it works in TS land but I always find this hard to read. Also at this point the timer would already have finished so the timer.stop() call executed by the cleanup is really not needed. Maybe the timer could have a pass/fail kind of thing. So you hook it up with to callbacks instead one that is when the timer has finished and one if it finishes early by the call to stop. The unbind would still be an issue with timeouts but if it had listeners it could be avoided but then the timer would be a lot more complex.

Something like this:
const timer = Timer(() => reject(...), resolve, cleanup)

Then timer.stop() would call the resolve a timeout would call reject and cleanup would be called on both those actions.

Reply:
Yeah, I admit I wasn't overly happy about this... however we can't do that because we have 3 different paths to a resolution here. We have:
1. The runnable function completes, in which case it should ignore the timer, `resolve` immediately and cancel the timer.
2. The runnable function throws an error, in which case it should ignore the timer, `reject` immediately and cancel the timer.
3. The timer finishes, in which case the runnable function result is ignored and `reject` immediately.

I'll have more of a think tomorrow and see if I can come up with something cleaner.

Reply:
Okay, I think that should be a lot nicer now as we're no longer calling things before they are defined. How does that look @spocke?